### PR TITLE
Feat 218 save and restore keymap

### PR DESCRIPTION
**Pull request by adamrocker**
Enable to save/restore the current device keymap. 
Keymap data is stored in Firebase at /keymaps/v1/saved-keymaps/{id}.
The data structure is as follow:

```
type SavedKeymap = {
  id?: string; // this entity's id
  author_uid: string; //  auth.uid
  title: string;
  desc: string;
  label_lang: KeyboardLabelLang;
  layout_options: { option: string; optionChoice: string }[] | null;
  keycodes: { [pos: string]: number }[];
  created_at?: Date;
  updated_at?: Date;
  vendor_id: number;
  product_id: number;
  product_name: number;
};
```

If the connected keyboard is registered and approved to Remap, the vendor_id/product_id/product_name are from its KeyboardDefinitionDocument. Otherwise they are from device information.

The **title** and **desc** are shown in the Popup to recognize what map it is.

スクリーンショット 2021-03-16 18 45 53
スクリーンショット 2021-03-16 18 46 15

### CODE REVIEW

**Review by yoichiro — CHANGES_REQUESTED** [review]
@adamrocker Thank you for implementing this great feature!

When I tested this feature, I noticed that there is no validation for filling in the title and the description. If a long description is filled in, the description was not rendered normally. I think that we need to introduce some validation rule. For example, the length of the title should be less than 255, and the length of the description should be less than 512 and applies the word wrap style. How about this?

スクリーンショット 2021-03-16 22 15 51

**Review by yoichiro — CHANGES_REQUESTED** [review]
@adamrocker BTW, if some keyboard definition document in Firestore and it has one or more related keymap documents and the definition is deleted, should the keymap documents be also deleted?

**Comment by adamrocker** [comment]
@yoichiro 
BTW, if some keyboard definition document in Firestore and it has one or more related keymap documents and the definition is deleted, should the keymap documents be also deleted?
I think, that should be better but ...

**Comment by adamrocker** [comment]
> I think that we need to introduce some validation rule. For example, the length of the title should be less than 255, and the length of the description should be less than 512 and applies the word wrap style. 

@yoichiro I set the const values to control the text length here.　https://github.com/remap-keys/remap/pull/406/files#diff-9cbafadd4a222bb964cab219655198372c7408c8c78ddcfdd458e9f7ab4c0c54R30-R31

**Comment by yoichiro** [comment]
> @yoichiro
> 
> > BTW, if some keyboard definition document in Firestore and it has one or more related keymap documents and the definition is deleted, should the keymap documents be also deleted?
> 
> I think, that should be better but ...

Yeah, actually, I guess that there are some cases which we should avoid to delete saved keymaps, and we can delete them later asynchronously. We can decide the way to delete them as a new issue.

**Comment by yoichiro** [comment]
> > I think that we need to introduce some validation rule. For example, the length of the title should be less than 255, and the length of the description should be less than 512 and applies the word wrap style.
> 
> @yoichiro I set the const values to control the text length here.　https://github.com/remap-keys/remap/pull/406/files#diff-9cbafadd4a222bb964cab219655198372c7408c8c78ddcfdd458e9f7ab4c0c54R30-R31

LGTM!

File: src/actions/storage.action.ts [hunk 1]
```diff
@@ -490,4 +515,183 @@ export const storageActionsThunk = {
...
+    );
+
+    if (resultList.success) {
+      dispatch(
+        StorageActions.updateSavedRegisteredKeymapList(
+          resultList.savedKeymaps as SavedRegisteredKeymapData[]
+        )
+      );
+    } else {
+      console.error(resultList.cause);
```
Comment (yoichiro) [inline src/actions/storage.action.ts:539]: @adamrocker In this case, it does not only output a debug log with `console.log`, but it should also show the error notification (and to send a error report via Error Reporting API).

File: src/actions/storage.action.ts [hunk 2]
```diff
@@ -490,4 +515,183 @@ export const storageActionsThunk = {
...
+    const resultList = await storage.instance!.fetchUnregisteredKeymaps(info);
+
+    if (resultList.success) {
+      dispatch(
+        StorageActions.updateSavedUnregisteredKeymapDataList(
+          resultList.savedKeymaps as SavedUnregisteredKeymapData[]
+        )
+      );
+    } else {
+      console.error(resultList.cause);
```
Comment (yoichiro) [inline src/actions/storage.action.ts:623]: @adamrocker Could you add the logic to show the error notification?

**Review by yoichiro — CHANGES_REQUESTED** [review]
@adamrocker I posted some comments, and please confirm them.

**Comment by yoichiro** [comment]
@adamrocker Thank you for many changes! I finished reviewing this pull request. I also created some commits to fix some issues. So, I think that this pull request looks good to me, but I want to test this diff a bit. Could you give me a few hours for the testing by me?

**Review by yoichiro — APPROVED** [review]
@adamrocker I tested some use cases, and I could confirmed this worked as I expected. LGTM!

**Comment by adamrocker** [comment]
@yoichiro We discussed about the type case and got the results as below:
- snake case in DB
- snake case in App with DB's data, however the name of type MUST have 'Data' at the end
- camel case in App

According to above rule, I changed the name of properties this commit https://github.com/remap-keys/remap/pull/406/commits/e8cf84a1e95e209cf97f29579ea1be801310e661

**Comment by yoichiro** [comment]
@adamrocker LGTM!

**Pull request merged**